### PR TITLE
haskell.compiler.ghc{901,921,HEAD}: fix aarch64-darwin build

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

- fix aarch64-darwin builds of haskell.compiler.ghc9{0,2}1 by applying the `autoSignDarwinBinariesHook ` change from `ghc/8.10.7.nix`

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/ghc/9.2.1.nix [hunk 1]
```diff
@@ -185,6 +187,9 @@ stdenv.mkDerivation (rec {
     export NIX_LDFLAGS+=" -rpath $out/lib/ghc-${version}"
   '' + lib.optionalString stdenv.isDarwin ''
     export NIX_LDFLAGS+=" -no_dtrace_dof"
+
+    # GHC tries the host xattr /usr/bin/xattr by default which fails since it expects python to be 2.7
+    export XATTR=${lib.getBin xattr}/bin/xattr
```
Comment (sternenseemann) [inline pkgs/development/compilers/ghc/9.2.1.nix:192]: This is incorrect, the `XATTR` env var is not yet supported by GHC's 9.2 branch, see https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6862

File: pkgs/development/compilers/ghc/9.2.1.nix [hunk 2]
```diff
@@ -242,12 +247,10 @@ stdenv.mkDerivation (rec {
...
     perl autoconf automake m4 python3
     ghc bootPkgs.alex bootPkgs.happy bootPkgs.hscolour
+  ] ++ lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
+    autoSignDarwinBinariesHook
   ] ++ lib.optionals enableDocs [
     sphinx
-  ] ++ lib.optionals stdenv.isDarwin [
-    # TODO(@sternenseemann): backport addition of XATTR env var like
-    # https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6447
-    xattr
```
Comment (sternenseemann) [inline pkgs/development/compilers/ghc/9.2.1.nix:250]: See above.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

**Comment by prusnak** [comment]
I updated the PR to contain only `autoSignDarwinBinariesHook` change.

**Review by sternenseemann — APPROVED** [review]

**Comment by prusnak** [comment]
the same issue is with the `haskell.compiler.ghc901` so I pushed another commit which fixes that as well

**Comment by sternenseemann** [comment]
Don't think you need to bother with GHC 9.0.1 for the moment as it doesn't support aarch64-darwin codegen AFAIK?

**Comment by prusnak** [comment]
> Don't think you need to bother with GHC 9.0.1 for the moment as it doesn't support aarch64-darwin codegen AFAIK?

Still, I guess it's better to not ship binaries that are killed instantly?

**Comment by prusnak** [comment]
Added another fix for `haskell.compiler.ghcHEAD`

**Comment by prusnak** [comment]
I am wondering whether we should include the following change to fix ghcHEAD build on aarch64-darwin:

``` patch
--- a/pkgs/top-level/haskell-packages.nix
+++ b/pkgs/top-level/haskell-packages.nix
@@ -128,7 +128,7 @@ in {
       llvmPackages = pkgs.llvmPackages_10;
     };
     ghcHEAD = callPackage ../development/compilers/ghc/head.nix {
-      bootPkgs = packages.ghc901; # no binary yet
+      bootPkgs = packages.ghc921; # no binary yet
       inherit (buildPackages.python3Packages) sphinx;
       # Need to use apple's patched xattr until
       # https://github.com/xattr/xattr/issues/44 and
```

Is there a reason why stick to `ghc901`? @sternenseemann

**Comment by sternenseemann** [comment]
@prusnak You'll have to try, but it *should* work. Reason 9.0.1 was picked was that it was the only GHC version fulfilling the “at most two major releases older” requirement for bootstrapping.

**Comment by prusnak** [comment]
> it _should_ work.

I am going to try it out

**Comment by prusnak** [comment]
Unfortunately ghcHEAD fails with

```
libraries/binary/src/Data/Binary/Put.hs:103:1: error:
    Data.Binary.FloatCast: Can't be safely imported!
    The module itself isn't safe.
    |
103 | import Data.Binary.FloatCast (floatToWord, doubleToWord)
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by prusnak** [comment]
> Unfortunately ghcHEAD build with ghc921 fails with the following:

I reverted the change and kept only autoSignDarwinBinariesHook change for ghcHEAD, so this PR can be merged.

**Pull request merged**